### ISSUE TITLE
joinQuery Should NOT throw in find

### ISSUE THREAD

**Issue by MarcGodard** [DaddyWarbucks/feathers-fletching#32]
joinQuery is throwing an error when there are no results on a find call. This should be optional.

**Comment by MarcGodard** [comment]
The first work around:

```
options = {
  customers_phones: {
    service: 'api/customers-phones',
    foreignKey: 'id',
    targetKey: 'customerId'
  },
  customers_emails: {
    service: 'api/customers-emails',
    foreignKey: 'id',
    targetKey: 'customerId'
  }
}

const joinQueriesHook = (options) => (context) => {
  if (context.method === 'find') {
    try {
      return joinQuery(options)(context)
    } catch (error) {
      context.result = {
        total: 0,
        data: []
      }
      return context
    }
  }
  return joinQuery(options)(context)
}

...

find: [joinQueriesHook(options)],
```

**Comment by MarcGodard** [comment]
If you need something more custom (or you can modify--not tested):

```
const specialQuery = () => {
  return async context => {
    let query = {
      paginate: false,
       query: {
         $select: ['specialId']
       }
    }
    query = object.assign(query, context.params.query.customers_special})
    delete context.params.query.customers_special
    const special = await context.app
      .service('api/customers-special')
      .find(query)
    const ids = special.filter((d) => d.specialId).map((d) => d.specialId)
    if (ids.length) {
      context.params.query.$and = [{ id: { $in: ids } }]
    }
    return context
  }
}
```

**Comment by fratzinger** [comment]
What Marc said. I was thinking that for a `find` the thrown error should be optional for a while, too!

**Comment by fratzinger** [comment]
I wrote a util for this purpose: https://github.com/fratzinger/feathers-utils/blob/main/src/utils/setResultEmpty.ts
I would love to integrate this into `joinQuery`. What do you think @DaddyWarbucks?

**Comment by DaddyWarbucks** [comment]
For sure. It definitely should not throw on `find` but should throw on all other methods. For find, it should do as you guys suggested and return an empty result. I remember thinking about this as I wrote it but can't remember why I ended up having it throw on `find`...? Probably just an oversight on my part. I would really like to write a v2 to this hook. There are a number of performance improvements that also need to be made, as well as supporting more operators. For example, the following should probably also work
```js
const query = {
  $or: [
    { 'artist.name': 'Johnny' },
    { 'artist.name': 'June' }
  ]
}

const query = {
  $or: [
    {
      $and:
        [
          {...whatever},
          {...whatever},
        ]
    }
  ]
}
```
It may get tricky, but any combination or nested combination of `$or` and `$and` should work. I have used https://www.npmjs.com/package/traverse a number of times in my own apps. And while I hate to introduce a new dependency, we need something that can recursively traverse/modify the query. Do you guys have any suggestions on this? Maybe we could write our own too...we basically know what 99% of queries will look like.

We can also make some improvements to how the "sub queries" manage their pagination. I have thought quite a bit about this and we can do better. As long as the parent query is not using `$skip`, then we can manage pagination of the sub queries better.

I will try to prioritize this over the holidays. I am using it more and more myself too.

**Comment by DaddyWarbucks** [comment]
Released in version `1.20.4`
